dali_tf_plugin/dali_dataset_op.cc

```
@@ -210,7 +270,189 @@ class DALIDatasetOp::Dataset::Iterator : public DatasetIterator<Dataset> {
+  }
+
+#if TF_MAJOR_VERSION == 2 && TF_MINOR_VERSION >= 3
+  Status SaveInternal(SerializationContext *ctx, IteratorStateWriter *writer) override {
+    return errors::Unimplemented("SaveInternal is not supported for DALI dataset.");
+  }
+
+  Status RestoreInternal(IteratorContext *ctx, IteratorStateReader *reader) override {
+    return errors::Unimplemented("RestoreInternal is not supported for DALI dataset");
+  }
+#endif
+
+ private:
+  // TODO(klecki): Check this again
+  // TensorFlow treats a sample/example as a vector of Tensors (they flatten everything),
+  // and if a dataset has multiple outputs it means, that it returned a tuple that maps
+  // to a vector of Tensors.
+  using TfExample = std::vector<Tensor>;
+
+  // Batch is a list of Tensors - we repack the TfExamples into a Batch.
+  // We need to build batches sample by sample to advance the input iterators in sync.
+  // TODO(klecki): In batch mode this will be a single Tensor with additional dimension representing
+  //               whole batch
+  using Batch = std::vector<Tensor>;
+
+  // Represents tuple of inputs for one iteration
+  using ListOfBatches = std::vector<Batch>;
+
+
+  /**
+   * @brief Schedule required number of runs of DALI Pipeline to fill the prefetch queue.
+   *
+   * When there are input datasets, feed the pipeline required number of input batches.
+   *
+   * TODO(klecki): Inputs handled only for an uniform executor
+   */
+  Status PrefetchPipeline(IteratorContext *context, daliPipelineHandle *pipeline_handle) {
+    if (!dataset()->pipeline_def_.exec_separated) {
+      int prefetch_depth = dataset()->pipeline_def_.prefetch_queue_depth;
+      if (dataset()->HasInputs()) {
+        for (int i = 0; i < prefetch_depth; i++) {
+          bool end_of_sequence;
+          ListOfBatches batches;
+          TF_RETURN_IF_ERROR(PrepareBatches(context, batches, &end_of_sequence));
+          if (end_of_sequence) {
+            return errors::InvalidArgument(
```

Can't we just silently ignore if the queue is bigger than data available?

---

Hmm, if we manage to get at least one iteration, we could probably just process that one iteration. I'm not sure which one is better. Should we be strict with the requested prefetching?